src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -231,35 +231,25 @@ public Task<DocumentOnAutoInsertResponseItem[]> GetDocumentOnAutoInsertAsync(Doc
             => _requestHandlerProvider.ExecuteRequestAsync<DocumentOnAutoInsertParams, DocumentOnAutoInsertResponseItem[]>(MSLSPMethods.OnAutoInsertName,
                 autoInsertParams, _clientCapabilities, _clientName, cancellationToken);
 
-#pragma warning disable VSTHRD100 // Avoid async void methods
-        private async void DiagnosticService_DiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
-#pragma warning restore VSTHRD100 // Avoid async void methods
+        private void DiagnosticService_DiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
         {
-            // Since this is an async void method, exceptions here will crash the host VS. We catch exceptions here to make sure that we don't crash the host since
-            // the worst outcome here is that guests may not see all diagnostics.
-            try
+            // LSP doesnt support diagnostics without a document. So if we get project level diagnostics without a document, ignore them.
+            if (e.DocumentId != null && e.Solution != null)
             {
-                // LSP doesnt support diagnostics without a document. So if we get project level diagnostics without a document, ignore them.
-                if (e.DocumentId != null && e.Solution != null)
+                var document = e.Solution.GetDocument(e.DocumentId);
+                if (document == null || document.FilePath == null)
                 {
-                    var document = e.Solution.GetDocument(e.DocumentId);
-                    if (document == null || document.FilePath == null)
-                    {
-                        return;
-                    }
-
-                    // Only publish document diagnostics for the languages this provider supports.
-                    if (document.Project.Language != CodeAnalysis.LanguageNames.CSharp && document.Project.Language != CodeAnalysis.LanguageNames.VisualBasic)
-                    {
-                        return;
-                    }
-
-                    // LSP does not currently support publishing diagnostics incrememntally, so we re-publish all diagnostics.
-                    await PublishDiagnosticsAsync(document).ConfigureAwait(false);
+                    return;
                 }
-            }
-            catch (Exception ex) when (FatalError.ReportWithoutCrash(ex))
-            {
+
+                // Only publish document diagnostics for the languages this provider supports.
+                if (document.Project.Language != CodeAnalysis.LanguageNames.CSharp && document.Project.Language != CodeAnalysis.LanguageNames.VisualBasic)
+                {
+                    return;
+                }
+
+                // LSP does not currently support publishing diagnostics incrememntally, so we re-publish all diagnostics.
+                Task.Run(() => PublishDiagnosticsAsync(document));
```

Comment:
How does this deal with exceptions? Should this be `PublishDiagnosticsAsync(document)..ReportNonFatalErrorAsync()` so that they're not lost?

Reply:
That is the subject of #29273